cmd/blockchaincmd/change_weight.go

```diff
@@ -141,59 +134,133 @@ func setWeight(_ *cobra.Command, args []string) error {
-		network,
-		blockchainName,
-		nodeID,
-		0, // automatic uptime
-		isBootstrapValidatorForNetwork(nodeID, sc.Networks[network.Name()]),
-		false, // don't force
-	)
+	if rpcURL == "" {
+		rpcURL, _, err = contract.GetBlockchainEndpoints(
+			app,
+			network,
+			chainSpec,
+			true,
+			false,
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	validationID, err := validatorManagerSDK.GetValidationID(rpcURL, nodeID)
 	if err != nil {
 		return err
 	}
 
-	// TODO: we need to wait for the balance from the removed validator to arrive in changeAddr
-	// set arbitrary time.sleep here?
-
-	weight, err = app.Prompt.CaptureWeight("What weight would you like to assign to the validator?")
+	vdrInfo, err := validatorManagerSDK.GetValidatorInfo(network.SDKNetwork(), validationID)
 	if err != nil {
 		return err
 	}
 
-	balance, err = getValidatorBalanceFromPChain()
+	totalWeight, err := validatorManagerSDK.GetTotalWeight(network.SDKNetwork(), subnetID)
 	if err != nil {
 		return err
 	}
 
-	publicKey, pop, err = getBLSInfoFromPChain()
+	allowedChange := float64(totalWeight) * constants.MaxL1TotalWeightChange
+
+	if float64(vdrInfo.Weight) > allowedChange {
+		return fmt.Errorf("can't make change: current validator weight %d exceeds max allowed weight change of %d", vdrInfo.Weight, uint64(allowedChange))
+	}
+
+	allowedChange = float64(totalWeight-vdrInfo.Weight) * constants.MaxL1TotalWeightChange
+
+	if newWeight == 0 {
+		ux.Logger.PrintToUser("Current validator weight is %d", vdrInfo.Weight)
+		newWeight, err = app.Prompt.CaptureWeight(
+			"What weight would you like to assign to the validator?",
+			func(v uint64) error {
+				if v > uint64(allowedChange) {
+					return fmt.Errorf("weight exceeds max allowed weight change of %d", uint64(allowedChange))
+				}
+				return nil
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	if float64(newWeight) > allowedChange {
+		return fmt.Errorf("can't make change: desired validator weight %d exceeds max allowed weight change of %d", newWeight, uint64(allowedChange))
+	}
+
+	publicKey, err = formatting.Encode(formatting.HexNC, bls.PublicKeyToCompressedBytes(vdrInfo.PublicKey))
 	if err != nil {
 		return err
 	}
 
-	remainingBalanceOwnerAddr, err = getChangeAddrFromPChain()
-	if err != nil {
-		return fmt.Errorf("failure parsing change owner address: %w", err)
+	if pop == "" {
+		_, pop, err = promptProofOfPossession(false, true)
+		if err != nil {
+			return err
+		}
 	}
 
-	// add back validator to subnet with updated weight
-	return CallAddValidator(deployer, network, kc, blockchainName, nodeID.String(), publicKey, pop)
-}
+	deployer := subnet.NewPublicDeployer(app, kc, network)
 
-// getValidatorBalanceFromPChain gets remaining balance of validator from p chain
-func getValidatorBalanceFromPChain() (uint64, error) {
-	return 0, nil
-}
+	var remainingBalanceOwnerAddr, disableOwnerAddr string
+	hrp := key.GetHRP(network.ID)
+	if len(vdrInfo.RemainingBalanceOwner.Addrs) > 0 {
```

Comment:
should we check if `vdrInfo.RemainingBalanceOwner` is not `nil` to avoid possible crash